File: superstitioMain/java/superstitio/relics/f_boss/LewdGirlComing.java
package superstitio.relics.f_boss;

import com.megacrit.cardcrawl.actions.common.HealAction;
import com.megacrit.cardcrawl.actions.common.RelicAboveCreatureAction;
import com.megacrit.cardcrawl.core.AbstractCreature;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.monsters.AbstractMonster;
import superstitio.DataManager;
import superstitio.powers.AbstractSuperstitioPower;
import superstitio.relics.SuperstitioRelic;
import superstitioapi.powers.interfaces.OnMonsterDeath;
import superstitioapi.utils.CreatureUtility;

public class LewdGirlComing extends SuperstitioRelic {
    public static final String ID = DataManager.MakeTextID(LewdGirlComing.class);
    public static final int DEATH_DOOR_AMOUNT = 1;
    // 遗物类型
    private static final RelicTier RELIC_TIER = RelicTier.BOSS;
    // 点击音效
    private static final LandingSound LANDING_SOUND = LandingSound.MAGICAL;
//    private static final int DrawAmount = 1;
//    private static final int EjaculationMax = 2;

    public LewdGirlComing() {
        super(ID, RELIC_TIER, LANDING_SOUND);
    }

//    @Override
//    public AbstractCard.CardColor getRelicOwner() {
//        return DataManager.SPTT_DATA.LupaEnums.LUPA_CARD;
//    }

    @Override
    public void atBattleStart() {
        AbstractMonster monster = CreatureUtility.getRandomMonsterSafe();
//        for (AbstractMonster monster : AbstractDungeon.getMonsters().monsters) {
        this.addToTop(new RelicAboveCreatureAction(monster, this));
        monster.addPower(new RatherThanDeath(monster, DEATH_DOOR_AMOUNT));
//        }
        AbstractDungeon.onModifyPower();
    }

//    @Override
//    public void onSpawnMonster(AbstractMonster monster) {
//        monster.addPower(new RatherThanDeath(monster, DEATH_DOOR_AMOUNT));
//        AbstractDungeon.onModifyPower();
//    }

    public void onEquip() {
        ++AbstractDungeon.player.energy.energyMaster;
    }

    public void onUnequip() {
        --AbstractDungeon.player.energy.energyMaster;
    }

    @Override
    public void updateDescriptionArgs() {
        setDescriptionArgs(DEATH_DOOR_AMOUNT);
    }

    public static class RatherThanDeath extends AbstractSuperstitioPower implements OnMonsterDeath.OnMonsterDeathPower {
        public static final String POWER_ID = DataManager.MakeTextID(RatherThanDeath.class);

        public RatherThanDeath(final AbstractCreature owner, int amount) {
            super(POWER_ID, owner, amount);
        }

        @Override
        public boolean ifStopOwnerDeathWhenOwnerIsMonster() {
            this.flash();
            addToTop_AutoRemoveOne(this);
            addToTop(new HealAction(this.owner, this.owner, DEATH_DOOR_AMOUNT));
            return true;
        }

        @Override
        public void updateDescriptionArgs() {
            setDescriptionArgs(this.amount);
        }

        @Override
        public void onDeath() {
            this.flash();
            addToBot(new HealAction(this.owner, this.owner, DEATH_DOOR_AMOUNT));
            addToBot_AutoRemoveOne(this);
        }
    }
}

File: superstitioMain/java/superstitio/relics/f_boss/LewdGirlComing.kt
package superstitio.relics.f_boss

import com.megacrit.cardcrawl.actions.common.HealAction
import com.megacrit.cardcrawl.actions.common.RelicAboveCreatureAction
import com.megacrit.cardcrawl.core.AbstractCreature
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import superstitio.DataManager
import superstitio.powers.AbstractSuperstitioPower
import superstitio.relics.SuperstitioRelic
import superstitio.relics.f_boss.LewdGirlComing.RatherThanDeath
import superstitioapi.powers.interfaces.OnMonsterDeath.OnMonsterDeathPower
import superstitioapi.utils.CreatureUtility
import superstitioapi.utils.setDescriptionArgs

class LewdGirlComing  //    private static final int DrawAmount = 1;
//    private static final int EjaculationMax = 2;
    : SuperstitioRelic(ID, RELIC_TIER, LANDING_SOUND)
{
    //    @Override
    //    public AbstractCard.CardColor getRelicOwner() {
    //        return DataManager.SPTT_DATA.LupaEnums.LUPA_CARD;
    //    }
    override fun atBattleStart()
    {
        val monster = CreatureUtility.getRandomMonsterSafe()
        //        for (AbstractMonster monster : AbstractDungeon.getMonsters().monsters) {
        this.addToTop(RelicAboveCreatureAction(monster, this))
        monster.addPower(RatherThanDeath(monster, DEATH_DOOR_AMOUNT))
        //        }
        AbstractDungeon.onModifyPower()
    }

    //    @Override
    //    public void onSpawnMonster(AbstractMonster monster) {
    //        monster.addPower(new RatherThanDeath(monster, DEATH_DOOR_AMOUNT));
    //        AbstractDungeon.onModifyPower();
    //    }
    override fun onEquip()
    {
        ++AbstractDungeon.player.energy.energyMaster
    }

    override fun onUnequip()
    {
        --AbstractDungeon.player.energy.energyMaster
    }

    override fun updateDescriptionArgs()
    {
        setDescriptionArgs(DEATH_DOOR_AMOUNT)
    }

    class RatherThanDeath(owner: AbstractCreature, amount: Int) : AbstractSuperstitioPower(POWER_ID, owner, amount),
        OnMonsterDeathPower
    {
        override fun ifStopOwnerDeathWhenOwnerIsMonster(): Boolean
        {
            this.flash()
            addToTop_AutoRemoveOne(this)
            addToTop(HealAction(this.owner, this.owner, DEATH_DOOR_AMOUNT))
            return true
        }

        override fun updateDescriptionArgs()
        {
            setDescriptionArgs(this.amount)
        }

        override fun onDeath()
        {
            this.flash()
            addToBot(HealAction(this.owner, this.owner, DEATH_DOOR_AMOUNT))
            addToBot_AutoRemoveOne(this)
        }

        companion object
        {
            val POWER_ID: String = DataManager.MakeTextID(RatherThanDeath::class.java)
        }
    }

    companion object
    {
        val ID: String = DataManager.MakeTextID(LewdGirlComing::class.java)
        const val DEATH_DOOR_AMOUNT: Int = 1

        // 遗物类型
        private val RELIC_TIER = RelicTier.BOSS

        // 点击音效
        private val LANDING_SOUND = LandingSound.MAGICAL
    }
}